Make the agency sign-up select dropdown searchable

wilmersondasilva opened the pull request:
This select dropdown did not allow the users to type ahead. 

#### Proposed Changes

* This PR introduces a solution for this using the react-select library.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to agency/signup
* Use the Country select, and try to search for some country
* Select some country that has available states, like the United States or Brazil, and then try to use the State select to search for some state.

#### Screenshots

![image](https://user-images.githubusercontent.com/6406071/190467463-ebd6cf62-5798-4b46-b23a-b948760360d4.png)

![image](https://user-images.githubusercontent.com/6406071/190467556-11f4cfdb-5f22-4cae-b918-1a2296e596de.png)

![image](https://user-images.githubusercontent.com/6406071/190467653-a36ebc13-2612-4871-916e-856e0ac73bc7.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

wilmersondasilva committed:
Rebase with Trunk - Make the agency sign-up select dropdown searchable

This select dropdown did not allow the users typeahead. This commit introduces a solution for this using the react-select library.

wilmersondasilva committed:
Agency Sign-Up: Make the state select dropdown searchable

This select dropdown did not allow the users typeahead. This commit changes it to be a searchable dropdown.

wilmersondasilva committed:
Agency Sign-Up: fix incorrect merge

The last merge contained the wrong stuff. This commit removes the unneeded lines of code.

kallehauge commented on client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx:
```diff
@@ -134,23 +158,38 @@ export default function CompanyDetailsForm( {
 							options={ countryOptions }
-							onSelect={ ( option: any ) => {
-								setCountry( option.value );
+							isDisabled={ isLoading }
+							isLoading={ countryOptions.length === 0 }
```
I'm not sure if the "disabled/loading" state is working as expecting. Personally I would expect the field to be greyed out or something similar while it's being loaded but we get a `- Select Country -` value in the field instead like the field is empty:

![Screenshot 2022-10-07 at 15 34 08](https://user-images.githubusercontent.com/3846700/194567190-23bc6d0d-18d3-407c-ae73-e7e1c874e5b3.jpg)

And then, when it's done loading, the fields with the existing values are shown:

![Screenshot 2022-10-07 at 15 34 15](https://user-images.githubusercontent.com/3846700/194567275-393e98aa-afb7-4f92-98ad-5df7c07f53b5.jpg)

wilmersondasilva replied:
You're right. The reason is the `countryOptions.length` will never be 0, because it'll always have at least one option, the `- Select Country -` option. I let it this way because it was what we had before. I just fixed it, and now the select input will be disabled while it loads the list of countries and will also show a loading indicator on its right.

![image](https://user-images.githubusercontent.com/6406071/198339855-c1822367-34a6-4079-9e2b-75bd6f03ca71.png)

wilmersondasilva replied:
You're right. The reason for this is the `countryOptions.length` will never 0, since it's has always the option `- Select Country -`. I just fixed the code to deal with this.

<img width="1254" alt="image" src="https://user-images.githubusercontent.com/6406071/198341573-25dd2bca-89ad-42dc-bdd5-8124c2f0d02e.png">

This is what the users will see while the countries are fetched.

![image](https://user-images.githubusercontent.com/6406071/198342652-b0d0cc88-888d-44ee-9a63-b62e05f7ba51.png)

This was the behavior expected, but it wasn't working because of the bug I found. It's relevant to notice that the list of countries is fetched only once and then cached.

kallehauge commented on client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx:
```diff
@@ -0,0 +1,55 @@
+		color: 'green',
+		padding: '0 8px 0 4px',
+	} ),
+	singleValue: ( styles ) => ( { ...styles, color: 'var(--color-neutral-70)', fontWeight: 600 } ),
```
The value of the select field seems to be displayed in a bolder weight (at least with my Chrome v. 105.0.5195.125):

![Screenshot 2022-10-07 at 15 41 40](https://user-images.githubusercontent.com/3846700/194567996-29f733ae-0b31-4589-9f19-3ef2561d8b32.jpg)

wilmersondasilva replied:
@kallehauge, I purposely let it as it was before. I didn't want to change it since I presumed it was approved by someone in the past. This is a screenshot of how it looks in the `trunk` branch.

![image](https://user-images.githubusercontent.com/6406071/198070488-267ec2b6-2ed4-4211-8dda-1c680cf82385.png)

kallehauge reviewed (changes requested):
Awesome first take!

I've got a couple of lights comments and it seems like we get an error in our `Unit tests (Web app)` check saying: `Process exited with code 1 (Step: Prevent duplicated packages)`. Perhaps the `react-select` library already exists?

wilmersondasilva committed:
Agency Sign-Up: fix a bug in the company-details-form

A bug made the country fields appear even though the country list didn't be available. It happened because the logic was to show the country fields when the length of the country list was different from 0. The issue here is the country list will always have at least one item, the default item. This commit fixed this by showing the country fields only when the country list length is

robertf4 commented:
I found a potential in-house alternative to `react-select` made by VIP. I asked [here](pdOhbm-bY-p2#comment-83) if we can use it in Calypso. If so then that might be preferable to pulling in a new third party library, and it seems that VIP has handled A11Y better than `react-select`.

djalmaaraujo commented:
> I found a potential in-house alternative to `react-select` made by VIP. I asked [here](pdOhbm-bY-p2#comment-83) if we can use it in Calypso. If so then that might be preferable to pulling in a new third party library, and it seems that VIP has handled A11Y better than `react-select`.

@robertf4 I replied to your question. Another alternative if you don't want to mix up the libraries because of the dependencies is using directly the https://alphagov.github.io/accessible-autocomplete/ component we used as the base component.

This is probably the best component for autocomplete regarding accessibility and usability.

If you don't want to include the vip-design-system library, you can look at our implementation https://github.com/Automattic/vip-design-system/blob/trunk/src/system/NewForm/FormAutocomplete.js.

Happy to help!

DustinHartzler commented:
I tried testing this and got an error message loading the `/agency/signup/` page:
![bgMekm2tPx.png](https://user-images.githubusercontent.com/2293077/203371540-54c55b7d-e100-4504-aae9-36b88bec6167.png)

It looks like the `react-select` library isn't loading properly.

DustinHartzler commented:
I needed to run `yarn install` first, whoops!

Works as expected. 

One improvement I would suggest is to clear out the field when it becomes focused, so it's more apparent you can type to search.

![CleanShot 2022-11-30 at 11 53 46](https://user-images.githubusercontent.com/2293077/204859532-d59a23bb-e643-42e7-905a-61ddd13f4e05.gif)

DustinHartzler committed:
Changed placeholder text

DustinHartzler committed:
Added CSS for placeholder text

DustinHartzler commented:
Updated the placeholder text of the country field as well as matching the placeholder colors of the Business Address field. 

![image](https://user-images.githubusercontent.com/2293077/206530262-e0ad33ca-5c2b-4dd3-86a7-be1fa232ad99.png)


## Testing Instructions

- Go to http://jetpack.cloud.localhost:3000/agency/signup in an incognito window
- You'll need to create a new WordPress.com test account or use a test account that has not signed up for an agency before
- Use the Country select, and try to search for some country
- Select some country that has available states, like the United States or Brazil, and then try to use the State select to search for some state.

yashwin commented on client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx:
```diff
@@ -4,24 +4,46 @@ import { ReactChild, useCallback, useState } from 'react';
+	return options[ 0 ];
+}
+
+function getState( state: string, options: StateOption[] ): StateOption {
```
How about we replace this with the [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) method?

yashwin reviewed (changes requested):
I had a look at this PR, and it works as expected. 

I left some comments and found a few UI/UX issues.

1) The border and highlight colors used here is blue in color, which I think should be black in color like other input boxes.

<img width="571" alt="Screenshot 2022-12-12 at 6 40 27 PM" src="https://user-images.githubusercontent.com/10586875/207055070-63e4a67d-4117-4d42-9877-593a95985451.png">

**Other input** 

<img width="609" alt="Screenshot 2022-12-12 at 6 55 25 PM" src="https://user-images.githubusercontent.com/10586875/207056088-34d38df5-6f21-493b-b07f-75a5eb492fa3.png">

2) When searching for an item in the country field, the text `Type to find country` is considered a dropdown item that also shows/hides based on the search term.

**Visible**

<img width="589" alt="Screenshot 2022-12-12 at 6 38 13 PM" src="https://user-images.githubusercontent.com/10586875/207055207-3a944d1c-19d8-4d2b-ba53-3734af04933d.png">

**Hidden**

<img width="588" alt="Screenshot 2022-12-12 at 6 38 25 PM" src="https://user-images.githubusercontent.com/10586875/207055959-49f2da86-163d-4f79-aa17-4717d986841d.png">


Can we also rebase this PR with the latest trunk?

DustinHartzler committed:
Updated drop down colors

DustinHartzler committed:
Fixed a couple TS errors

yashwin committed:
Minor fixes and a few refactor to the country and state selection code

tyxla commented on client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx:
```diff
@@ -0,0 +1,68 @@
+import { Gridicon } from '@automattic/components';
+import Select, { StylesConfig, DropdownIndicatorProps, components } from 'react-select';
```
I'd discourage introducing yet another third-party component for this, given that we have our own alternatives.

Unless I'm missing something, we should be able to use `ComboboxControl` from `@wordpress/components` - see [component docs](https://developer.wordpress.org/block-editor/reference-guides/components/combobox-control/) for more info.

This component is already in use in a couple of places of Calypso:

https://github.com/Automattic/wp-calypso/blob/349a6b6a0fe384efcc8651ce47f0db4f4bf46fa1/client/landing/stepper/declarative-flow/internals/steps-repository/store-address/index.tsx#L246

https://github.com/Automattic/wp-calypso/blob/349a6b6a0fe384efcc8651ce47f0db4f4bf46fa1/client/signup/steps/woocommerce-install/step-store-address/index.tsx#L164

So I don't see a good reason to use another library for the same purpose. If there's a feature missing, we still might prefer to contribute upstream to support that feature, instead of using another 3rd-party library.

Here's some context as to why `@wordpress/components` are preferred:

p3N5Fy-uu-p2

p58i-7Ei-p2

Hope that helps, but please let me know if you run into any trouble or if you have any further questions.

jkguidaven replied:
@tyxla I tried to use the `ComboboxControl`, but it doesn't seem to support the `disabled` props. In case we want to contribute by adding the missing feature. How does that work on the timing of release? Do we immediately get the changes in Calypso, or will we have to wait for the next Gutenberg release?

jkguidaven replied:
Ohh I just found out. It looks like there is a weekly release.

tyxla replied:
I also wonder if wrapping the `ComboboxControl` in a `<Disabled />` component will do the job for you.

jkguidaven replied:
@tyxla that works. 👍 I'll use that instead. Thanks.



I was looking at `inert`, but react doesn't play nice with it and makes the implementation hacky.

```
ref={ ( node ) => node && ( disabled ? node.setAttribute( 'inert', '' ) : node.removeAttribute( 'inert' ) )
```

tyxla replied:
Glad to hear it worked for you 👍

DustinHartzler committed:
Started refactoring in ComboboxControl

DustinHartzler committed:
Added styling to ComboBoxControl fields

DustinHartzler committed:
Fixed country field to use ComboBoxControl

DustinHartzler committed:
Added State field Logic

jkguidaven committed:
Remove custom searchable dropdown component.

jkguidaven committed:
Code cleanup.

jkguidaven committed:
Add disable state.

jkguidaven committed:
Fix default state value.

jkguidaven committed:
Remove is-loading state.

jkguidaven commented on client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:
```diff
@@ -0,0 +1,61 @@
+.searchable-dropdown {
+	.components-base-control__field.components-base-control__field {
```
You may wonder why this is so. Currently, an inline class is injected into the inner component with styling that always prevails.

Instead of using `!important` , this is another approach to force override.

tyxla commented on client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx:
```diff
@@ -0,0 +1,28 @@
+			{ ! disabled && <ComboboxControl { ...props } /> }
+
+			{ disabled && (
+				<Disabled>
```
If `Disabled` supports a `isDisabled` prop that you can use by passing `disabled` to it. That way you won't have to render two separate `ComboboxControl` instances.

jkguidaven replied:
Though this work but the type definition didn't include the `isDisabled` props, which makes CI complain that it doesn't exist, 🤔 I'm not sure if there is something we can do about it.

```
import { ComponentType, Consumer, ReactNode } from 'react';

declare namespace Disabled {
    interface Props {
        children: ReactNode;
    }
}
declare const Disabled: {
    (props: Disabled.Props): JSX.Element;
    Consumer: Consumer<boolean>;
};

export default Disabled;
```

tyxla replied:
Well, the type definition includes `isDisabled` in trunk: https://github.com/WordPress/gutenberg/blob/b439e5f1a305f1c49d0817772aa0e7b02fa6c863/packages/components/src/disabled/types.ts#L7

It was added 9 months ago in #42708. 

The `@wordpress/components` package now ships its own types, so when we upgrade to the new version of `@wordpress/components`, it will be able to benefit from those types. We're currently working on the upgrade in #77121 and we're hoping to conclude it in the following weeks.

In the meantime, feel free to add a temporary type patch or even a `@ts-ignore` and a `TODO` comment there, pointing out that upgrading the `@wordpress/components` package to use its own types will fix it.

jkguidaven replied:
Thanks for looking into this. Good to hear this will be fixed soon. I added the `ts-ignore` as suggested. 👍

tyxla commented on client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:
```diff
@@ -0,0 +1,61 @@
+}
+
+.searchable-dropdown.is-disabled {
+	pointer-events: none;
```
This should be coming by default from the `Disabled` component.

jkguidaven replied:
Good catch 👍 this has been fixed.

tyxla commented on client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:
```diff
@@ -0,0 +1,61 @@
+		border-radius: 2px;
+		background-color: var(--color-surface);
+		transition: all 0.15s ease-in-out;
+		height: 40px;
```
It's a real bummer we have to override so many styles. 😞 

I wonder if some of them can be centrally applied to properly adapt and style the combobox control better for the entirety of Calypso 🤔

jkguidaven replied:
I agree with this. I would say we can maybe do this on a separate PR. A couple of areas in which the combo box is being used have similar styling. We can probably decide what would be the standard Calypso styling from there. wdyt?

tyxla replied:
Sounds good, no need to address it here. Perhaps we can add a TODO comment then to remember to address it.

jkguidaven replied:
This has been added. 👍

tyxla commented on client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts:
```diff
@@ -119,5 +114,5 @@ export function useCountriesAndStates() {
 			stateOptionsMap: stateOptions,
 		};
-	}, [ countriesList ] );
+	}, [ countriesList, translate ] );
```
👍

tyxla reviewed (commented):
Thanks for adapting it to use `@wordpress/components` 👍 

I've left some suggestions on the code and components side of things. Let me know what you think

jkguidaven committed:
Use 'isDisabled' props on Disabled component and remove pointer-event CSS styling.

jkguidaven committed:
Revert 'isDisabled' props due to type definition error.

jkguidaven committed:
Use isDisabled props in Disabled component with @ts-ignore to avoid compiler from complaining.

jkguidaven committed:
Add Todo comment for applying standard Calypso styling for the combobox control.

TheSteveK reviewed (approved):
It looks like the code has been well and truly picked through and refined.  Tested and LGTM so 🚢

yashwin reviewed (approved):
I have tested the changes, and it LGTM. Good job on this, @jkguidaven! 🚀

tyxla reviewed (approved):
Nothing further from me, either, feel free to go ahead 🚀

jkguidaven committed:
Fix rebase issue.

jkguidaven commented:
@yashwin @TheSteveK @tyxla thank you for reviewing this PR 🙇

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8026009

Thank you @wilmersondasilva for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8026009).
